packages/@sanity/schema/src/sanity/validation/types/array.ts

```diff
@@ -106,6 +133,79 @@ export default (typeDef, visitorContext) => {
     )
   }
 
+  const [primitiveTypes, objectTypes] = partition(
+    of,
+    (ofType) =>
+      isPrimitiveTypeName(ofType.type) ||
+      isPrimitiveTypeName(visitorContext.getType(ofType.type)?.jsonType)
+  )
+
+  const isMixedArray = primitiveTypes.length > 0 && objectTypes.length > 0
+
+  if (isMixedArray) {
+    const primitiveTypeNames = primitiveTypes.map((t) => t.type)
+    const objectTypeNames = objectTypes.map((t) => t.type)
+    problems.push(
+      error(
```

Comment:
Would prefer of the error-message creation code (here and below) was extracted into functions; will make it easier to follow the flow/improve readability imo.